data/evidence/store.go

```diff
@@ -0,0 +1,273 @@
+		true,
+		func(key, value []byte) bool {
+			ar, err := (&AllegationRequest{}).FromBytes(value)
+			if err != nil {
+				fmt.Println("failed to deserialize allegation request")
+				return false
+			}
+			return fn(ar)
+		},
+	)
+}
+
+func (es *EvidenceStore) CheckRequestExists(new keys.Address) bool {
+	requestAlreadyExists := false
+	es.IterateRequests(func(ar *AllegationRequest) bool {
+		if ar.MaliciousAddress.Equal(new) {
+			requestAlreadyExists = true
+		}
+		return false
+	})
+	return requestAlreadyExists
+}
+
+func (es *EvidenceStore) CreateSuspiciousValidator(validatorAddress keys.Address, status int8, height int64, createdAt *time.Time) (*LastValidatorHistory, error) {
+	lvh := NewLastValidatorHistory(validatorAddress, status, height, createdAt)
+	err := es.UpdateSuspiciousValidator(lvh)
+	return lvh, err
+}
+
+func (es *EvidenceStore) UpdateSuspiciousValidator(lvh *LastValidatorHistory) error {
+	dat, err := serialize.GetSerializer(serialize.PERSISTENT).Serialize(lvh)
+	if err != nil {
+		return err
+	}
+	return es.Set(es.getSuspiciousValidatorKey(lvh.Address), dat)
+}
+
+func (es *EvidenceStore) IsFrozenValidator(validatorAddress keys.Address) bool {
+	dat, err := es.Get(es.getSuspiciousValidatorKey(validatorAddress))
+	if err != nil {
+		return false
+	}
+	if len(dat) == 0 {
+		return false
+	}
+	lvh := &LastValidatorHistory{}
+	err = serialize.GetSerializer(serialize.PERSISTENT).Deserialize(dat, lvh)
+	if err != nil {
+		return false
+	}
+
+	if err != nil {
+		return false
+	}
+	return lvh.IsFrozen()
+}
+
+func (es *EvidenceStore) HandleRelease(options *Options, validatorAddress keys.Address, blockHeight int64, blockCreatedAt time.Time) error {
+	lvh, err := es.GetSuspiciousValidator(validatorAddress, 0, 0)
+	if err != nil {
+		return err
+	}
+	if !lvh.IsFrozen() {
+		return fmt.Errorf("Validator \"%s\" already released", validatorAddress)
+	}
+
+	isReady, err := lvh.ReleaseReady(options, blockCreatedAt)
+	if err != nil {
+		return err
+	}
+
+	if !isReady {
+		return fmt.Errorf("Validator \"%s\" not ready for release", validatorAddress)
+	}
+
+	lvh.ReleaseHeight = blockHeight
+	lvh.ReleaseAt = &blockCreatedAt
+	err = es.UpdateSuspiciousValidator(lvh)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (es *EvidenceStore) PerformAllegation(validatorAddress keys.Address, maliciousAddress keys.Address, ID string, blockHeight int64, proofMsg string) error {
+	es.mux.Lock()
+	defer es.mux.Unlock()
+
+	isBusy := es.IsRequestIDBusy(ID)
+	if isBusy {
+		return fmt.Errorf("request ID %s already handled\n", ID)
+	}
+
+	ar := NewAllegationRequest(ID, validatorAddress, maliciousAddress, blockHeight, proofMsg)
+	if es.CheckRequestExists(ar.MaliciousAddress) {
+		return ErrRequestAlreadyExists
+	}
+	err := es.SetAllegationRequest(ar)
+	if err != nil {
+		return err
+	}
+
+	at, err := es.GetAllegationTracker()
+	if err != nil {
+		return err
+	}
+
+	at.Requests[ID] = true
+
+	err = es.SetAllegationTracker(at)
```

Comment:
Can we sort the map inside 'at' before save?

Reply:
Hmm ,the map is being cleaned at every block 
CleanTracker() , does exactly that . 

Do you mean to call the same function here as well , for additional safety ?